// after-brunch/lib/index.js
// Generated by CoffeeScript 1.7.1
(function() {
  var AfterBrunch, exec, sysPath;

  sysPath = require('path');

  exec = require('child_process').exec;

  module.exports = AfterBrunch = (function() {
    AfterBrunch.prototype.brunchPlugin = true;

    function AfterBrunch(config) {
      var _ref;
      this.config = config;
      this.commands = (_ref = this.config.plugins.afterBrunch) != null ? _ref : [];
    }

    AfterBrunch.prototype.onCompile = function(generatedFiles) {
      var command, sf, _i, _len, _ref, _ref1;
      sf = (_ref = generatedFiles[0]) != null ? _ref.sourceFiles[0] : void 0;
      if (!(sf != null ? sf.path.match(/\.(css|styl)$/) : void 0)) {
        _ref1 = this.commands;
        for (_i = 0, _len = _ref1.length; _i < _len; _i++) {
          command = _ref1[_i];
          exec(command, function(error, stdout, stderr) {
            if (error) {
              return console.log(error.message);
            } else {
              if (stdout) {
                console.log(stdout);
              }
              if (stderr) {
                return console.log(stderr);
              }
            }
          });
        }
      }
    };

    return AfterBrunch;

  })();

}).call(this);
