// LiveScript-brunch/lib/index.js
// Generated by CoffeeScript 1.3.3
(function() {
  var LiveScript, LiveScriptCompiler, sysPath;

  LiveScript = require('LiveScript');

  sysPath = require('path');

  module.exports = LiveScriptCompiler = (function() {

    LiveScriptCompiler.prototype.brunchPlugin = true;

    LiveScriptCompiler.prototype.type = 'javascript';

    LiveScriptCompiler.prototype.extension = 'ls';

    function LiveScriptCompiler(config) {
      this.config = config;
      null;
    }

    LiveScriptCompiler.prototype.compile = function(data, path, callback) {
      var error, result;
      try {
        return result = LiveScript.compile(data, {
          bare: true
        });
      } catch (err) {
        return error = err;
      } finally {
        callback(error, result);
      }
    };

    //LiveScriptCompiler.prototype.include = [sysPath.join(__dirname, '..', 'vendor', 'prelude-browser-0.6.0.js')];
    LiveScriptCompiler.prototype.include = [sysPath.join(__dirname, '..', 'vendor', 'prelude-browser-min.js')];

    return LiveScriptCompiler;

  })();

}).call(this);
